addons/delivery/models/sale_order.py

```diff
@@ -37,7 +37,13 @@ def onchange_order_line(self):
             self.recompute_delivery_price = True
 
     def _remove_delivery_line(self):
-        self.env['sale.order.line'].search([('order_id', 'in', self.ids), ('is_delivery', '=', True)]).unlink()
+        delivery_line = self.env['sale.order.line'].search([('order_id', 'in', self.ids), ('is_delivery', '=', True)])
+        if not delivery_line:
+            return
+        to_delete = delivery_line.filtered(lambda x: x.qty_invoiced == 0)
+        if not to_delete:
+            raise UserError(_('You can not update the shipping costs on an order where it was already invoiced!'))
```

Comment:
The translation file (.pot) should be updated (cc @mart-e)

Reply:
And usually, it's good to provide a bit of details like the specific line (like "product x has been invoiced y")

Reply:
> And usually, it's good to provide a bit of details like the specific line (like "product x has been invoiced y")

Not sure, is the functionality still there where you can have more than one delivery line (created from a picking after confirmation)? If so, I guess we might have an additional issue with the all in one deletion anyway.

@mart-e 
How about a bot adding a properly generated .pot file automatically? 😉

Reply:
> How about a bot adding a properly generated .pot file automatically? wink

@wtaferner no because I don't want to encourage this ;-) when you are adding/modifying term, you are creating English content in a potentially fully translated db, something we should avoid when possible

Reply:
@mart-e 
Alright, I was not 100% convinced, but as you were asking I made it very explicit now, in case there is more than one delivery line which is invoiced and no other line to be deleted. Let's hope it won't be scrapped in the future and replaced with a non-explicit error message 😉 

About the bot, yeah, there are still some work to do to have clean po file generated, but it was not that bad and IMO in my local instance (some cleanup). Therefore it should be possible in a clean database without demo data to create in case detected the proper pot file via a bot and add it to a PR automatically. Anyway, unrelated to this PR.